topics/dev/tutorials/interactive-tools/tutorial.md

```diff
@@ -0,0 +1,872 @@
+>
+> ```sh
+> ansible-playbook galaxy.yml
+> ```
+>
+> {% endraw  %}
+>
+{: .tip}
+
+---
+
+# Debugging
+
+The most obvious way to test a tool is simply to run it in the Galaxy UI, straight from the tool panel. If you are extremely lucky, you will find that the tool starts up and runs without error. But we all know that never happens! So this is where we start iteratively debugging our tool, until it functions as expected.
+
+> ### {% icon comment %} A successful tool run
+> It is worth pointing out that the appearance of a GxIT in the Galaxy user history is not intuitive when you are used to running "regular" tool jobs. When the history item turns orange ("processing"), that's when a GxIT is actually ready to use! At this point, the tool UI should refresh and display a link to the active GxIT. Remember, the history item doesn't turn green until a job has terminated. With a GxIT, that only happens when the tool has been stopped by the user, or by wall time limits imposed by the Galaxy administrators.
+{: .comment}
+
+> ### {% icon tip %} GxIT testing infrastructure
+> Testing and debugging is currently the trickiest part of GxIT development. Ideally, Galaxy core will be developed in the future to better support the process, but for the time being we have to make the most with what is available! In the future, we would like to see GxITs being tested with Planemo, and being installed and tested by Ephemeris from the ToolShed.
+{: .tip}
+
+
+---
+
+# Additional components
+
+The GxIT that we wrapped in this tutorial was a simple example, and you should now understand what is required to create an Interactive Tool for Galaxy. However, there are a few additional components that can enhance the reliability and user experience of the tool. In addition, more complex applications may require some additional components or workarounds the create the desired experience for the user.
+
+## Self-destruct script
+Web server applications will tend to keep running after the user has terminated the tool in Galaxy, which can result in "zombie" containers hanging around and clogging up the Galaxy server. It is therefore good practice to implement a script that will end the server process when the connection to Galaxy has been terminated. The following script will watch Galaxy's binding to the container port and kill the container when the connection terminates (e.g. the user has ended the job). Just change `RUN_COMMAND_NAME` to the run command for your application.
+
+```sh
+RUN_COMMAND_NAME='my_run_script.sh'
+while true; do
+    sleep 60
+    if [ `netstat -t | grep -v CLOSE_WAIT | grep ':8888' | wc -l` -lt 1 ]; then
+        pkill $RUN_COMMAND_NAME
+    fi
+done
+```
+
+This script can then be built into the docker image and called in the tool XML `<command>` before the main application call. Alternatively, it can be called within a `run.sh` script inside the container itself (see below).
+
+```sh
+/scripts/monitor_traffic.sh &
+bash my_run_script.sh
+```
```

Comment:
This seems a bit outdated, this was true for GIEs but with GxITs that support the model of "leave it running in the background", we've mostly stopped doing this, I don't think we should recommend it here.

Reply:
Thanks Helena, I wasn't sure of the history of this script so great to know, thanks!